lib/vnet/install_service_windows.go

```diff
@@ -0,0 +1,165 @@
+	svc, err := svcMgr.OpenService(serviceName)
+	if err != nil {
+		// The service has not been created yet and must be installed.
+		svc, err = svcMgr.CreateService(
+			serviceName,
+			tshPath,
+			mgr.Config{
+				StartType: mgr.StartManual,
+			},
+			ServiceCommand,
+		)
+		if err != nil {
+			return trace.Wrap(err, "creating VNet Windows service")
+		}
+	}
+	if err := svc.Close(); err != nil {
+		return trace.Wrap(err, "closing VNet Windows service")
+	}
+	if err := grantServiceRights(); err != nil {
+		return trace.Wrap(err, "granting authenticated users permission to control the VNet Windows service")
+	}
+	return nil
+}
+
+// UninstallService uninstalls the VNet windows service.
+func UninstallService(ctx context.Context) (err error) {
+	svcMgr, err := mgr.Connect()
+	if err != nil {
+		return trace.Wrap(err, "connecting to Windows service manager")
+	}
+	svc, err := svcMgr.OpenService(serviceName)
+	if err != nil {
+		return trace.Wrap(err, "opening Windows service %s", serviceName)
+	}
+	if err := svc.Delete(); err != nil {
+		return trace.Wrap(err, "deleting Windows service %s", serviceName)
+	}
+	if err := svc.Close(); err != nil {
+		return trace.Wrap(err, "closing VNet Windows service")
+	}
+	return nil
+}
+
+func grantServiceRights() error {
+	// Get the current security info for the service, requesting only the DACL
+	// (discretionary access control list).
+	si, err := windows.GetNamedSecurityInfo(serviceName, windows.SE_SERVICE, windows.DACL_SECURITY_INFORMATION)
+	if err != nil {
+		return trace.Wrap(err, "getting current service security information")
+	}
+	// Get the DACL from the security info.
+	dacl, _ /*defaulted*/, err := si.DACL()
+	if err != nil {
+		return trace.Wrap(err, "getting current service DACL")
+	}
+	// Build an explicit access entry allowing authenticated users to start,
+	// stop, and query the service.
+	ea := []windows.EXPLICIT_ACCESS{{
+		AccessPermissions: windows.SERVICE_QUERY_STATUS | windows.SERVICE_START | windows.SERVICE_STOP,
+		AccessMode:        windows.GRANT_ACCESS,
+		Trustee: windows.TRUSTEE{
+			TrusteeForm:  windows.TRUSTEE_IS_NAME,
+			TrusteeType:  windows.TRUSTEE_IS_WELL_KNOWN_GROUP,
+			TrusteeValue: windows.TrusteeValueFromString("Authenticated Users"),
+		},
+	}}
+	// Merge the new explicit access entry with the existing DACL.
+	dacl, err = windows.ACLFromEntries(ea, dacl)
+	if err != nil {
+		return trace.Wrap(err, "merging service DACL entries")
+	}
+	// Set the DACL on the service security info.
+	if err := windows.SetNamedSecurityInfo(
+		serviceName,
+		windows.SE_SERVICE,
+		windows.DACL_SECURITY_INFORMATION,
+		nil,  // owner
+		nil,  // group
+		dacl, // dacl
+		nil,  // sacl
+	); err != nil {
+		return trace.Wrap(err, "setting service DACL")
+	}
+	return nil
+}
+
+// assertProgramFiles asserts that tsh is installed under the program files
+// directory (usually C:\Program Files\).
+func assertProgramFiles(tshPath string) error {
+	programFiles := os.Getenv("PROGRAMFILES")
```

Comment:
Could this value be user-controlled? Does that pose a problem?

Reply:
It could, but I don't think that's a problem, and I don't think there's any better way to find the appropriate installation directory. The user now needs admin rights to run the installer so even if they can change the env var it's no worse than anything else they can do.